Ingest structured ES server logs

ycombinator opened the pull request:
This PR teaches the `elasticsearch/server` fileset to ingest structured Elasticsearch server logs.

This PR takes the same approach as https://github.com/elastic/beats/pull/10352, in that it creates an entrypoint pipeline, `pipeline.json`, that delegates further processing of a log entry depending on what it sees as the first character of the entry:
- If the first character is `{`, it assumes the log line is structured as JSON and delegates further processing to `pipeline-json.json`.
- Else, it assumes the log line is plaintext and delegates further processing to `pipeline-plaintext.json`.

elasticmachine commented:
Pinging @elastic/stack-monitoring

ycombinator commented:
/cc @pgomulka for awareness

cachedout commented:
Just as a possible concern here, this looks like it's going to break if a log line contains valid JSON but does not begin with `{`. Understanding that it's not likely that the ES team will change this in the future, as a forward-looking goal it might be good to formalize this contract a little more.

ycombinator commented:
Until #10445 gets merged in, this PR will fail CI on the system tests for the `elasticsearch/deprecation` fileset with errors like this:

```
Exception: Key 'elasticsearch.server.component' found in event is not documented!
```

ycombinator commented:
@ruflin I just added a couple more sample log fixtures containing stack trace logs (as provided by @pgomulka in https://github.com/elastic/beats/pull/9988). I'd like you to take a look at the `field.yml` definition I added for this field, especially from an ECS POV, and the generated `test-json.log-expected.json` file (you will need to expand the file as it's too big). Thanks!

cachedout commented on filebeat/module/elasticsearch/server/manifest.yml:
```diff
@@ -4,10 +4,13 @@ var:
+      - /var/log/elasticsearch/*_server.json
     os.darwin:
       - /usr/local/var/lib/elasticsearch/*.log
+      - /usr/local/var/lib/elasticsearch/*_server.json
```
A quick question about these paths. For the unstructured logs it looks like we're grabbing `*.log` which I assume would also include the slow logs but then for the structured logs it looks like they're explicitly breaking the ES log out to be be suffixed with `*_server.json`. So it would appear from this that for unstructured logs we're slurping up more log files that we are for structured logs, no? Is this by intention?

ycombinator replied:
Good question. There's a couple pieces here that need to be tied together to see the full picture:

1. These variables here in the `manifest.yml` are just variables. As in, they don't (yet) determine any configuration.

2. The actual configuration is built from here: https://github.com/elastic/beats/blob/7777c1d8593b2e714468f9f831a7288af09141fb/filebeat/module/elasticsearch/server/config/log.yml#L1-L15

   As you can see, the `path` variable from (1) are used there but also notice the `exclude_files` setting. This is how we make sure we only slurp up structured logs.

cachedout replied:
Gotcha. Makes perfect sense. Thanks!

ruflin commented on filebeat/module/elasticsearch/server/_meta/fields.yml:
```diff
@@ -2,10 +2,12 @@
-  - name: component
-    description: "Log component"
-    example: "o.e.c.m.MetaDataCreateIndexService"
+  - name: stacktrace
```
I wonder if we should even index the stacktrace? Or only have it as part of the source. Or alternative is this one of the fields that actually should be `text` as I expect it to be pretty long?

ycombinator replied:
Perhaps this could be a boolean option for this fileset? Something like `include_stacktrace`?

ruflin replied:
That is an other option in case space becomes an issue. At the same time if there is an error, first thing I do is look at the stacktrace so I think it's good to have it there. If someone does not want it, he could still use a processor to drop the field.

So I rather have it in here, I'm more asking if text is the correct type.

ycombinator replied:
Gotcha, I think for now we should not index it. It will still be shown as part of the source. The user would be be able to use other fields to narrow down their search (e.g. `log.level:ERROR' and then view the stack trace.

ycombinator replied:
In other words, I don't think it will be too common to want to search for something in the stack trace itself.

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

ycombinator commented:
If you expand the `test-json.log-expected.json` file in this PR, you'll notice that there are no `@timestamp` field. This is **not actually what we expect**. However, this is a known issues introduced by a [recent regression](https://github.com/elastic/elasticsearch/issues/38345) in Elasticsearch. Ideally Elasticsearch would fix the regression. Failing that, we could [try to adapt](https://github.com/elastic/beats/pull/10543) Filebeat ingest pipelines to handle the effects of the regression while also being backwards compatible.

In the spirit of not blocking this PR, I've removed the `@timestamp` fields from the expected JSON file for now. This matches up with similar changes done for other `elasticsearch` module filesets currently in `master`.
